src/decode.rs

```diff
@@ -47,63 +41,102 @@ pub(crate) enum PushGeneric<'a> {
+                            |_| {
+                                Err(DecodeError {
+                                    index: offset + byte_index,
+                                    kind: DecodeErrorKind::InvalidHexDigit,
+                                    mat: s.to_string(),
+                                    //mat: string[byte_index..].to_string(),
+                                })
+                            },
+                            |x| Ok(x as u32),
+                        )?,
+                    )
+                }
+            }
             "\\u" => {
-                let hex6 = &mat.as_str().as_bytes()[2..];
-                debug_assert_eq!(6, hex6.len());
-                let d0 = from_hex2(&hex6[0..2]) as u32;
-                let d1 = from_hex2(&hex6[2..4]) as u32;
-                let d2 = from_hex2(&hex6[4..]) as u32;
-
-                let c32 = (d0 << 16) + (d1 << 8) + d2;
-                match char::from_u32(c32) {
-                    Some(c) => {
-                        // It is a valid UTF code point. Always
-                        // decode it as such.
-                        let mut out = String::with_capacity(4);
-                        out.push(c);
-                        push_val(PushGeneric::String(&out))?;
-                        last_end = mat.end();
-                        continue;
-                    },
-                    // It is not a valid code point. Still try
-                    // to record it's value "as is".
-                    None => c32,
+                if rest < 8 {
+                    Err(DecodeError {
+                        index: offset + byte_index,
+                        kind: DecodeErrorKind::HexNumberToShort,
+                        mat: string[byte_index..].to_string(),
+                    })?
+                } else {
+                    let hex6 = &string[(byte_index + 2)..byte_index + 8].as_bytes();
+                    debug_assert_eq!(6, hex6.len());
+                    let c32 = from_hex6(hex6).map_err(|_| DecodeError {
+                        index: offset + byte_index,
+                        kind: DecodeErrorKind::InvalidHexDigit,
+                        mat: string[byte_index..].to_string(),
+                    })?;
+                    match char::from_u32(c32) {
+                        Some(c) => {
+                            // It is a valid UTF code point. Always
+                            // decode it as such.
+                            let mut out = String::with_capacity(4);
+                            out.push(c);
+                            push_val(PushGeneric::String(&out))?;
+                            string = &string[(byte_index + 8)..];
+                            start_index = 0;
+                            offset += byte_index + 8;
+                            continue;
+                        }
+                        // It is not a valid code point. Still try
```

Comment:
This code (which was originally mine!) is confusing. Please just do as the others do -- `PushGeneric::Value` followed by `continue`.

I _think_ if you do that you can remove _all_ of the `continue` statements.